Objects/regions.c

```diff
@@ -670,8 +720,277 @@ PyObject* _Py_MakeImmutable(PyObject* obj)
+
+    // The object is already known and doesn't need to be traversed
+    if (Py_REGION(obj) == info->region || _Py_IsImmutable(obj)) {
+        return 0;
+    }
+
+    // The actual region update and write checks are done in the
+    // main body of `add_to_region`
+    if (stack_push(info->pending, obj)) {
+        PyErr_NoMemory();
+        return -1;
+    }
+
+    return 0;
+}
+
+/* This adds the given object and transitive objects to the given region.
+ * Errors will be passed to the given `handle_error` function along with
+ * the `handle_error_data`.
+ *
+ * The given region may contain flags. Objects from the local regions will
+ * have their region replaced with the `flagged_region` value. This will
+ * lose flags set on the local region.
+ */
+PyObject *add_to_region(PyObject *obj, Py_uintptr_t flagged_region,
+                        handle_add_to_region_error handle_error,
+                        void* handle_error_data)
+{
+    if (!obj) {
+        Py_RETURN_NONE;
+    }
+
+    // Make sure we check against the actual region and not the region
+    // plus magic flags
+    Py_uintptr_t region = flagged_region & Py_REGION_MASK;
+
+    // The current implementation assumes region is a valid pointer. This
+    // restriction can be lifted if needed
+    assert(!IS_DEFAULT_REGION(region) || !IS_IMMUTABLE_REGION(region));
+
+    // Early return if the object is already in the region or immutable
+    if (Py_REGION(obj) == region || _Py_IsImmutable(obj)) {
+        Py_RETURN_NONE;
+    }
+
+    addtoregionvisitinfo info = {
+        .pending = stack_new(),
+        .region = region,
+    };
+    if (info.pending == NULL) {
+        return PyErr_NoMemory();
+    }
+
+    if (stack_push(info.pending, obj)) {
+        stack_free(info.pending);
+        return PyErr_NoMemory();
+    }
+
+    regionmetadata *region_data = _Py_CAST(regionmetadata *, region);
+    while (!stack_empty(info.pending)) {
+        PyObject *item = stack_pop(info.pending);
+        PyTypeObject *type = Py_TYPE(item);
+        regionmetadata *item_region = _Py_CAST(regionmetadata *, Py_REGION(item));
+
+        if (Py_REGION(item) == region) {
+            // -1 for the refernce we just followed
+            region_data->lrc -= 1;
+            continue;
+        }
+
+        if (IS_DEFAULT_REGION(Py_REGION(item))) {
+            // Add reference to the object,
+            // minus one for the reference we just followed
+            region_data->lrc += item->ob_refcnt - 1;
+            Py_SET_REGION(item, flagged_region);
+            item_region = region_data;
+
+            if (PyFunction_Check(item)) {
+                regionerror err = {.src = _PyObject_CAST(region_data->bridge),
+                    .tgt = item, .id = ERR_WIP_FUNCTIONS };
+                if ((handle_error)(&err, handle_error_data)) {
+                    continue;
+                } else {
+                    goto fail;
+                }
```

Comment:
Is this a temporary error as we add more features or the intended long term behaviour?  Could you add a comment either way please.

Reply:
I guess WIP says that but being clearer would be good.